chains/evm/contracts/pools/AdvancedPoolHooks.sol

```diff
@@ -0,0 +1,236 @@
+
+import {EnumerableSet} from "@openzeppelin/contracts@5.0.2/utils/structs/EnumerableSet.sol";
+
+/// @notice Advanced pool hooks for additional security features like allowlists and CCV management.
+/// @dev This is a standalone contract that can optionally be used by TokenPools.
+contract AdvancedPoolHooks is IAdvancedPoolHooks, Ownable2StepMsgSender {
+  using EnumerableSet for EnumerableSet.AddressSet;
+
+  error AllowListNotEnabled();
+  error SenderNotAllowed(address sender);
+
+  event AllowListAdd(address sender);
+  event AllowListRemove(address sender);
+  event CCVConfigUpdated(
+    uint64 indexed remoteChainSelector,
+    address[] outboundCCVs,
+    address[] outboundCCVsToAddAboveThreshold,
+    address[] inboundCCVs,
+    address[] inboundCCVsToAddAboveThreshold
+  );
+  event ThresholdAmountSet(uint256 thresholdAmount);
+
+  struct CCVConfig {
+    address[] outboundCCVs; // CCVs required for outgoing messages to the remote chain.
+    address[] outboundCCVsToAddAboveThreshold; // Additional CCVs that are required for outgoing messages above threshold to the remote chain.
+    address[] inboundCCVs; // CCVs required for incoming messages from the remote chain.
+    address[] inboundCCVsToAddAboveThreshold; // Additional CCVs that are required for incoming messages above threshold from the remote chain.
+  }
+
+  struct CCVConfigArg {
+    uint64 remoteChainSelector;
+    address[] outboundCCVs;
+    address[] outboundCCVsToAddAboveThreshold;
+    address[] inboundCCVs;
+    address[] inboundCCVsToAddAboveThreshold;
+  }
+
+  /// @dev The immutable flag that indicates if the allowlist is access-controlled.
+  bool internal immutable i_allowlistEnabled;
+
+  /// @dev A set of addresses allowed to trigger lockOrBurn as original senders.
+  /// Only takes effect if i_allowlistEnabled is true.
+  /// This can be used to ensure only token-issuer specified addresses can move tokens.
+  EnumerableSet.AddressSet internal s_allowlist;
+
+  /// @dev Threshold token transfer amount above which additional CCVs are required.
+  /// Value of 0 means that there is no threshold and additional CCVs are not required for any transfer amount.
+  uint256 internal s_thresholdAmountForAdditionalCCVs;
+
+  /// @dev Stores verifier (CCV) requirements keyed by remote chain selector.
+  mapping(uint64 remoteChainSelector => CCVConfig ccvConfig) internal s_verifierConfig;
+
+  constructor(address[] memory allowlist, uint256 thresholdAmountForAdditionalCCVs) {
+    // Allowlist can be set as enabled or disabled at deployment time only to save hot-path gas.
+    i_allowlistEnabled = allowlist.length > 0;
+    if (i_allowlistEnabled) {
+      _applyAllowListUpdates(new address[](0), allowlist);
+    }
+    s_thresholdAmountForAdditionalCCVs = thresholdAmountForAdditionalCCVs;
+  }
+
+  /// @inheritdoc IAdvancedPoolHooks
+  function preflightCheck(Pool.LockOrBurnInV1 calldata lockOrBurnIn, uint16, bytes calldata) external view {
```

Comment:
This implementation only does an allowlist check. Other hook contracts could be written that do things like ACE